ct/l1: plumbing for Raft-backed metastore STM

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
This PR adds plumbing and boilerplate code as a precursor for a metastore implementation that is backed by Raft-backed partitions of the L1 metastore topic. Namely, this PR:
- introduces a persisted STM for the L1 metastore, with mostly trivial calls into the existing `state` class, reusing existing code from `simple_metastore`
- introduces a "domain manager" that wraps the STM (the spiritual equivalent of datalake::coordinator or the NTP archiver)
- manages domain managers with the domain supervisor on leaders of the metastore topic partitions
- passes through requests from the frontend to the domain managers

Conceptually, all the plumbing (frontend, domain manager, supervisor, STM) is now there to implement the metastore interface with the STMs. This PR stops short of actually implementing it though, since the amount of code just for this plumbing is pretty non-trivial.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
ct/l1: add simple_stm

This adds a simple Raft-backed state machine that updates an in-memory
cloud_topics::l1::state instance via state_updates.

This is entirely a copy-paste of coordinator/state_machine.*, but with
the appropriate types swapped in.

This does not yet include other components for making this useful, like:
- a frontend to route requests to the leader
- a metastore implementation that wraps calls to this STM-backed state
  (similar to simple_stm)

andrwng committed:
ct/l1: introduce domain manager

Introduces the domain_manager, an abstraction that will manage access to
the L1 domains, each backed by a partition of the L1 metastore topic.
To this end, each domain manager wraps an STM owned by the partition.

The STM used currently is the simple_stm. This allows the the domain
manager to reuse some code from simple_metastore to serve requests, but
with a state backed by a Raft group rather than one entirely in memory.

The direct caller of the domain manager will be the frontend, as the
endpoint of RPC calls. As such, the domain manager exposes of interface
with rpc:: request/reply types.

andrwng committed:
ct/l1: expose static methods from simple_metastore

The logic implemented in simple_metastore is going to be reused by the
replicated version of simple_metastore. To that end, this pulls the
logic out into static methods.

andrwng committed:
ct/l1: misc field updates to rpc_types

Adds a few missing fields and error codes that will be useful to route
metastore requests. These were mostly an oversight.

Also introduces a state_error enum that will be the return code for a
logical state machine error.

andrwng committed:
ct/l1: manage domain_managers from domain_supervisor

In the same spirit of datalake::coordinator_manager managing
datalake::coordinators, this makes domain_supervisor manage
domain_managers, using the new partition notification interface to
ensure that each metastore partition leader instantiates a domain
manager.

andrwng committed:
ct/l1: plumb domain supervisor into L1 frontend

Since the L1 frontend needs to route to the appropriate domain manager,
this updates the frontend to us the domain supervisor to route to the
appropriate domain manager, if it exists.

andrwng committed:
ct/l1: unify L1 topic creation in domain supervisor

Previously the frontend and the domain_supervisor each had roughly
equivalent code to create the L1 metastore topic. Now that the frontend
has access to the domain supervisor, this PR updates it to call into the
supervisor to create the metastore topic.

Along the way, this updates the behavior to create more than one
partition, as was done by the frontend previously.

andrwng committed:
application: construct services for L1 metastore

Instantiates the L1 metastore frontend and the RPC service that wraps it
in the Redpanda application, and adds the STM factory that instantiates
the STM for the Raft members of the L1 topic partitions.

This will allow us to implement a metastore client implementation that
interacts with the L1 metastore topic partition leaders (will be in a
later commit).

andrwng committed:
cluster: expose raft manager

This will be useful to construct a partition_change_notifier from just
the controller, which I'll do from the domain supervisor in a later
commit.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b0d46a1b-4911-4191-9e1e-455032137831-->
<!--TEST RESULTS-->
<!--af5c18d9-66e0-4ede-8a44-5e81f80d4dce-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70228">build#70228</a></summary>


| test_class                | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70228#01987b4f-6bb8-4540-8bfc-049b56e71c4a | FLAKY         | 15/21    | upstream reliability is '93.8584779706275'. current run reliability is '71.42857142857143'. drift is 22.42991 and the allowed drift is set to 50. The test should PASS  |
| FeaturesMultiNodeTest     | test_license_upload_and_query | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70228#01987b51-cb76-4ee5-8b54-1d625e7f6632 | FLAKY         | 20/21    | upstream reliability is '92.36641221374046'. current run reliability is '95.23809523809523'. drift is -2.87168 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70265">build#70265</a></summary>


| test_class      | test_method                 | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:----------------|:----------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| MasterTestSuite | test_replica_pair_frequency |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/70265#01987cec-72fe-4594-b73a-9f5b6093a7be | FAIL          | 0/1      |          |
</details>
<!--/body-->

rockwotj commented on src/v/cloud_topics/level_one/metastore/rpc_types.h:
```diff
@@ -157,9 +164,12 @@ struct get_compaction_offsets_request
+    }
 
     model::topic_id_partition tp;
+    model::timestamp tombstone_removal_upper_bound_ts;
```
a comment for what this means? Not obvious to me.

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/metastore/simple_metastore.h:
```diff
@@ -47,6 +47,16 @@ class simple_object_builder : public metastore::object_metadata_builder {
+      const state&, const model::topic_id_partition&, model::timestamp);
+    static std::expected<compaction_offsets_response, errc>
+    get_compaction_offsets(
+      const state&, const model::topic_id_partition&, model::timestamp);
```
should these be private + friend?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/metastore/simple_stm.h:
```diff
@@ -0,0 +1,106 @@
+// persisted and replicated, making it slightly more useful for end-to-end
+// testing that may involve multiple nodes.
+//
+// This is not a full solution though, since it keeps everyhing in memory. It
```
```suggestion
// This is not a full solution though, since it keeps everything in memory. It
```

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/domain/domain_manager.cc:
```diff
@@ -0,0 +1,254 @@
+    builder.add_raw_kv(
+      serde::to_iobuf(add_objects_update::key),
+      serde::to_iobuf(std::move(update_res.value())));
+    auto repl_res = co_await stm_->replicate_and_wait(
+      sync_res.value(), std::move(builder).build(), as_);
+    if (!repl_res.has_value()) {
+        co_return rpc::add_objects_reply{
+          .ec = convert_stm_errc(repl_res.error()),
+        };
+    }
```
Is it important that we return `rpc::errc::concurrent_requests` like we do in the above case? It's not clear to me why it's a soft guarantee that this operation was actually applied.

andrwng replied:
Conceptually I think this shouldn't always be `rpc::errc::concurrent_requests`. If for instance this replica lost leadership when replicating, `rpc::errc::concurrent_requests` isn't an appropriate response, and perhaps the same request but routed to the new leader would be an appropriate course of action.

andrwng replied:
Or maybe you're referring to the success case? Fair, if there was an apply error, we can detect that by checking for the objects. Let me add that in

andrwng replied:
Done, but lmk if I've misunderstood what you're asking about

rockwotj replied:
> Or maybe you're referring to the success case? Fair, if there was an apply error, we can detect that by checking for the objects. Let me add that in

This is what I was referring to, although good point on the error code. I still think it's not quite bulletproof, as another (valid) update could be applied between the scheduling point returning from replication and this code running. But yeah I think it's fine if this over reports, because the user should retry from scratch and detect a noop.

rockwotj commented on src/v/cloud_topics/level_one/domain/domain_supervisor.cc:
```diff
@@ -26,23 +30,57 @@ namespace experimental::cloud_topics::l1 {
+              const model::ntp& ntp,
+              std::optional<cluster::partition_change_notifier::partition_state>
+                partition) {
+                // NOTE: listen on all partition notification types.
```
shouldn't we filter for at least L1 domain topic + namespace?

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/domain/domain_supervisor.cc:
```diff
@@ -26,23 +30,57 @@ namespace experimental::cloud_topics::l1 {
             _as.request_abort();
             co_await *std::exchange(_loop, std::nullopt);
         }
+        co_await _queue.shutdown();
+        _partition_notifications->unregister_partition_notifications(
+          _partition_notifications_id);
```
does it make sense to stop notifications, then shutdown the queue?

andrwng replied:
I'm not sure it matters, but maybe @bharathv has thoughts about this. If we get a notification in between these lines, the queue submit will fail, which is fine.

FWIW I copied the order from datalake_manager, which has the additional twist that notifications can resubmit to the queue if they fail, without seeing another notification. I omitted that in this PR since resetting domain managers is simpler than resetting translators, given all the scheduling around the translators.

bharathv replied:
Feels like the natural order is to stop notifications first and then shutdown queue (so nothing is enqueued after shutdown) but I think it doesn't matter because work_queue::submit(fn) is a no-op once shutdown is initiated.

> which has the additional twist that notifications can resubmit to the queue if they fail, without seeing another notification

yes this code is unnecessary complicated, we could get rid of those resubmits I think. 

Unrelated, shouldn't this stop() call `stop_and_wait()` on each remaining domain (from _domains) after `queue.shutdown()` for a proper cleanup?

andrwng replied:
Done and done

rockwotj commented on src/v/cloud_topics/level_one/domain/domain_supervisor.cc:
```diff
@@ -26,23 +30,57 @@ namespace experimental::cloud_topics::l1 {
+            _controller->get_partition_manager(),
+            _controller->get_topics_state()))
+      , _queue([](const std::exception_ptr& ex) {
+          vlog(cd_log.warn, "Unexpected domain supervisor error: {}", ex);
```
this should be an `error` log

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/domain/domain_supervisor.cc:
```diff
@@ -26,23 +30,57 @@ namespace experimental::cloud_topics::l1 {
     }
 
     ss::future<> stop() {
+        auto close_gate_fut = _gate.close();
```
nothing holds this gate open, and we have the queue, so I don't think this is needed?

andrwng replied:
Good catch, done

rockwotj commented on src/v/cloud_topics/level_one/domain/domain_supervisor.cc:
```diff
@@ -175,7 +213,66 @@ class domain_supervisor::impl {
+        partition) {
+        auto is_l1_topic = ntp.ns == model::kafka_internal_namespace
+                           && ntp.tp.topic == model::l1_metastore_topic;
+        if (_gate.is_closed() || !is_l1_topic) {
```
I think the `is_l1_topic` filter can go outside the queue.

The `_gate.is_closed()` check is not needed. The work_queue stops right after the current task it's processing completes (it does not drain the full queue).

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/metastore/frontend.h:
```diff
@@ -33,6 +33,10 @@ concept request_has_topic_id_partition = requires(T t) {
     { t.tp } -> std::same_as<model::topic_id_partition&>;
 };
 
+namespace experimental::cloud_topics {
+class app;
+} // namespace experimental::cloud_topics
```
Is this dependency right? This feels like we're going to end up with some kind of cycle with this...

rockwotj replied:
Feels like the right thing is for the supervisor to get plumbed directly into the frontend and not the app?

Because the app is going to eventually need to contain the frontend right because that will be how the read path accesses the metastore?

andrwng replied:
Yeah, this did feel a bit shaky. Let me see how plumbing in the supervisor works out

andrwng replied:
I'm good on plumbing the supervisor through to the frontend, but one thing I'm unsure of is the app "containing" the frontend. The app is sharded, but the frontend is also necessarily sharded so it can route calls to appropriate shards -- I think this makes the lifecycles a bit hard to couple together and makes me think they need to be separate, though I may be missing something obvious.

andrwng replied:
Done, but lmk if you're suggesting additional changes for the frontend being contained by app

rockwotj replied:
Yeah we need to probably rename `app`. Please think a bit here, I'm good with this but we should probably rename app, because app makes me think a single entry point

bharathv commented on src/v/cloud_topics/level_one/metastore/frontend.cc:
```diff
@@ -26,37 +28,71 @@ namespace experimental::cloud_topics::l1 {
-  cluster::partition_manager&, model::ntp, rpc::get_offsets_request) {
-    co_return rpc::get_offsets_reply{.ec = rpc::errc::ok};
+  experimental::cloud_topics::app& app,
+  model::ntp ntp,
```
nit: const ref ntp?

andrwng replied:
Done

bharathv reviewed (commented):
all raft/stm pieces lgtm, didn't focus too much on the core metastore changes.

andrwng committed:
ct/l1: add comment about metastore arg

Realized in review that this wasn't documented explicitly.

andrwng committed:
ct: expose domain supervisor in app

This will be useful to pass the supervisor to the frontend, which will
live outside of the app.

andrwng commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/080123a6400ac96d8aaf0c3c05bd0b84f27fb09d..2e60a337dfb837b25fdf615e56e20180224ac66c), main functional updates:
- rejiggered plumbing of domain_supervisor to the frontend
- updated domain_supervisor shutdown
- added check for successful apply to domain_manager

andrwng commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/2e60a337dfb837b25fdf615e56e20180224ac66c..91119a1980ed904c143c6d93c9474d0daca8e3de) to rebase + resolve the renamed `take_raft_snapshot()` conflict

rockwotj commented on src/v/cloud_topics/level_one/metastore/simple_stm.cc:
```diff
@@ -0,0 +1,223 @@
+                        .handle_exception([this](const std::exception_ptr& e) {
+                            vlog(_log.warn, "failed to write snapshot: {}", e);
+                        })
+                        .finally([holder = _gate.hold()] {});
```
Why `spawn_with_gate` and this `finally` block?

bharathv replied:
I think this block of code was inspired by dl/coordinator/stm.. the reason is handle_exception() accesses some `this` state (`_log`) and there is a scheduling point between `spawn_with_gate_then` and `handle_exception` we want the gate to be held until handle_exception is done.

andrwng replied:
Ah, copy-pasta! Will update if this PR ends up needing another push

andrwng replied:
Oops, raced. Thanks for the color Bharath.

rockwotj reviewed (approved):
LGTM

bharathv commented on src/v/cloud_topics/level_one/domain/domain_supervisor.cc:
```diff
@@ -43,6 +73,28 @@ class domain_supervisor::impl {
+          _partition_notifications_id);
+        co_await _queue.shutdown();
+        for (auto& [_, domain_mgr] : _domains) {
+            co_await domain_mgr->stop_and_wait();
```
nit: (Feel free to skip).. think stop_and_wait() could throw, so probably better to handle exceptions and invoke stop on all instead of throwing at first exception. (also parallel_for_each)

andrwng replied:
+1, will update if this PR needs another push. Otherwise I can do this in a followup

dotnwat reviewed (commented):
solid gold
